src/hooks/UseGlobalAuth.ts

```diff
@@ -0,0 +1,136 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+//@ts-ignore
+import { v4 as uuidv4 } from "uuid";
+import jwt_decode from "jwt-decode";
+import { useMemo, useState } from "react";
+import {
+  useDeskproAppClient,
+  useDeskproAppEvents,
+  useInitialisedDeskproAppClient,
+} from "@deskpro/app-sdk";
+
+import { ISettings } from "../types/settings";
+import { getAccessAndRefreshTokens } from "../api/preInstallApi";
+
+export const useGlobalAuth = () => {
+  const { client } = useDeskproAppClient();
+  const key = useMemo(() => uuidv4(), []);
+
+  const [callbackUrl, setCallbackUrl] = useState<string | null>(null);
+  const [poll, setPoll] = useState<(() => Promise<{ token: string }>) | null>(
+    null
+  );
+  const [settings, setSettings] = useState<ISettings | null>(null);
+  const [accessCode, setAccessCode] = useState<string | null>(null);
+  const [message, setMessage] = useState<{
+    error?: string;
+    success?: string;
+  } | null>(null);
+
+  useDeskproAppEvents(
+    {
+      onAdminSettingsChange: setSettings,
+    },
+    []
+  );
+
+  useInitialisedDeskproAppClient(
+    (client) => {
+      (async () => {
+        const { callbackUrl, poll } = await client
+          .oauth2()
+          .getAdminGenericCallbackUrl(
+            key,
+            /\?(code)=(?<token>[^&]+)/,
+            // eslint-disable-next-line no-useless-escape
+            /(&state)=(?<key>[^&]+)/
+          );
+
+        setCallbackUrl(callbackUrl);
+        setPoll(() => poll);
+      })();
+    },
+    [key]
+  );
+
+  const signOut = () => {
+    client?.setAdminSetting("");
+
+    setAccessCode(null);
+  };
+
+  const signIn = async () => {
+    if (!callbackUrl || !poll) {
+      setMessage({
+        error:
+          "Error getting callback URL. Please wait for the app to be initialized.",
+      });
+
+      return;
+    }
+
+    window.open(
```

Comment:
This will probably get pop-up blocked - it should be a standard synchronous link

Reply:
With just a <Link to={callbackUrl}> wrapper>?

Reply:
`<a href="..." target="_blank"></a>`

Reply:
grand